### DISCUSSION TITLE
Kit crashing when scrolling in clipboard history

### DISCUSSION THREAD

**Discussion by Alarid** [johnlindquist/kit#381] [Errors]
I have had this issue for quite a while now, so I thought it was worth investigating and asking for help. This happens with the "clipboard-history" script: it usually works fine, except when I scroll down through history "too much", then Kit crashes and I need to kill Kit process in order to be able to re-start it. 

Here is a video of the issue I'm describing:

https://user-images.githubusercontent.com/5678065/128183721-6c6db4cd-1a2e-4203-9566-981249438ae8.mov

Here is the script:

```js
// Menu: Clipboard History
// Description: Copy something from the clipboard history
// Shortcut: command shift v

let { history } = await db(kitPath("db", "clipboard-history.json"))

let { value, type } = await arg("What to paste?", () => {
  return history.map(({ value, type, timestamp, secret }) => {
    return {
      type,
      name: secret ? value.slice(0, 4).padEnd(10, "*") : value,
      value: {
        value,
        type,
      },
      description: timestamp,
      preview:
        type === "image"
          ? md(`![timestamp](${value})`)
          : value.includes("\n")
          ? `<div class="font-mono text-xs">${value
              .split("\n")
              .map((line) => `<p>${line}</p>`)
              .join("")}<div>`
          : null,
    }
  })
})

if (type === "image") {
  await copyPathAsImage(value)
  await keystroke("command v")
}

if (type === "text") {
  await setSelectedText(value)
}
```

I can't find anything relevant to the problem in the logs. Anyone else facing this issue? Is there a way to solve it ATM?

**Comment by johnlindquist** [comment]
My bet is that it's crashing on trying to render the preview of that specific clipboard item. Maybe it has html in it?

If you remove the "preview" property from the script, does it still crash?

**Reply by Alarid** [reply]
No, it doesn't crash anymore, thanks John for your quick reply 😁

A "cleaner" solution would be to strip out the HTML of the preview?

**Reply by johnlindquist** [reply]
Yeah, I'll work on the preview component. Thanks for reporting this 👍

**Comment by Alarid** [comment]
Hey @johnlindquist, it's been a year already and I have a new issue with the clipboard history script! I haven't changed it since I removed the preview, and it stopped working, maybe because of an update I missed or a new API?

Basically, when I run the script from the Kit App Prompt or using my custom shortcut, nothing happens, while it should show me the prompt and ask me which entry from the clipboard history I want to paste. I took a look at the logs, but nothing obvious there:

```
[2022-09-06 14:54:56.602] [info] 🏃‍♀️ Run /Users/yohann/.kenv/scripts/clipboard-history.js
[2022-09-06 14:54:56.606] [info] Found:
[2022-09-06 14:54:56.614] [info] 38073: 🏎 /Users/yohann/.kenv/scripts/clipboard-history.js
[2022-09-06 14:54:56.620] [info] 38116: 🟢 start idle Prompt
[2022-09-06 14:54:56.624] [info] ⚠️ ABANDON 38015
[2022-09-06 14:54:56.629] [info] DISCONNECT
[2022-09-06 14:54:56.629] [info] EXIT { pid: 38015, code: 0 }
[2022-09-06 14:54:56.629] [info] 38015: 🟡 exit 0. Prompt process: /Users/yohann/.kenv/scripts/clipboard-history.js
[2022-09-06 14:54:56.629] [info] 38015: 🛑 removed
[2022-09-06 14:54:56.635] [info] ↖ OPEN: { x: 2304, y: 187, width: 768, height: 480 }
[2022-09-06 14:54:57.688] [info] Hiding because blur
```

How could I debug this to find a way to fix it, please?

**Reply by johnlindquist** [reply]
Unfortunately, it's a bug with the internal clipboard history that snuck into that build.

I'm releasing a new build in the next couple of days that should address it. Really sorry about that 😔

**Reply by Alarid** [reply]
Ok no worries, thanks (again) for your quick reply

**Reply by johnlindquist** [reply]
@Alarid If you'd like to join our new Discord server, I've posted a preview build which fixes the clipboard history there

https://discord.gg/uPBu2Fg7

**Reply by Ambushfall** [reply]
Hey @Alarid,

I've been poking around the idea of clipboard history, and it honestly seemed pretty cool, so I attempted to reverse engineer the missing piece, the Copy part.

I'd like your input on this, or any tips


```js
// Menu: Copy to Clipboard
// Description: Save to Clipboard history
// Shortcut: command shift c

import "@johnlindquist/kit"

const { history, write } = await db("clipboard-history", { history: [{ value: "", type: "", timestamp: "", secret: "" }] })

const value = await clipboard.readText();


const newValue = {
    value: value,
    timestamp: new Date(Date.now()).toLocaleString('en-GB', { timeZone: 'UTC' }),
    secret: value.includes('secret'),
    type: /(http)?s?:?(\/\/[^"']*\.(?:png|jpg|jpeg|gif|png|svg))/i.test(value) ? "image" : "text"
}

history.push(newValue)

await write()

```

**Reply by Alarid** [reply]
@Ambushfall I'm not sure to follow. There was no missing piece, using the script I provided the clipboard content is automatically used, the idea is just to browse through the history and to be able to paste again an old element. If it isn't working, it might be because the worker associated with the clipboard hasn't started. You can start it from Kit's contextual menu (this didn't exist at the time)
